### ISSUE TITLE
T0: Disentangle Blocklevel rules preservation from workflow archival time 

### ISSUE THREAD

**Issue by todor-ivanov** [dmwm/WMCore#11042]
**Impact of the new feature**
WMAgent, T0

**Is your feature request related to a problem? Please describe.**
While working on the [issue](https://github.com/dmwm/WMCore/issues/10904) related to the T0 WMStats data retention policy, we figured out that (and we were also warned by T0 Team) that the postponing of the archival time by increasing [archivaldelay](https://github.com/dmwm/T0/blob/2abca2a5c8b0ccdc8bfbff340b48ce8302096387/bin/00_deploy_prod.sh#L193
) parameter to 6 months will also make the output data at EOS being kept for that long as well. 

The current Feature Request is to ensure we disentangle the real data preservation for T0 from the Metadata kept at T0 WMStats.  

**Describe the solution you'd like**
By looking at the code we believe we have figured out where the change needs  to happen. There is this [deleteBlocks](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py#L331) method in the `RucioInjectorPoller` which is used only by T0, and is basically constructing the set of block rules that are suitable for deletion. It is doing so by using the following sql [query](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/RucioInjector/Database/MySQL/GetDeletableBlocks.py#L27-L54):
```
    sql = """SELECT dbsbuffer_block.blockname,
                    dbsbuffer_location.pnn,
                    dbsbuffer_dataset.path,
                    dbsbuffer_dataset_subscription.site
             FROM dbsbuffer_dataset_subscription
             INNER JOIN dbsbuffer_dataset ON
               dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
             INNER JOIN dbsbuffer_block ON
               dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
             INNER JOIN dbsbuffer_file ON
               dbsbuffer_file.block_id = dbsbuffer_block.id
             INNER JOIN dbsbuffer_workflow ON
               dbsbuffer_workflow.id = dbsbuffer_file.workflow
             INNER JOIN dbsbuffer_location ON
               dbsbuffer_location.id = dbsbuffer_block.location
             LEFT OUTER JOIN wmbs_workflow ON
               wmbs_workflow.name = dbsbuffer_workflow.name
             WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
             AND dbsbuffer_dataset_subscription.subscribed = 1
             AND dbsbuffer_block.status = 'Closed'
             AND dbsbuffer_block.deleted = 0
             GROUP BY dbsbuffer_block.blockname,
                      dbsbuffer_location.pnn,
                      dbsbuffer_dataset.path,
                      dbsbuffer_dataset_subscription.site
             HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
             AND COUNT(wmbs_workflow.name) = 0
             """
```  

And the way how we can see the set of blocks returned could be related to workflows in `archived` status is by the very last line of the query, which is asking for all those block to have no records left in `WMBS` related to them. This clearing should be happening at archival time and driven by [TaskArchivaerPoller](https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/TaskArchiver/TaskArchiverPoller.py) and the `WMBS` record deletion itself is in [CleanCouchPoller](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/TaskArchiver/CleanCouchPoller.py#L379-L410).

So few possible solutions could happen here:
* Change the `sql` query so that we remove the last line requirement for not having records in `WMBS` connected to those blocks to be deleted - this to me sounds quite risky. We need to find out if that would make any dangerous deletions, and could cause data loss. But it would be contained in the code which is used only by T0 
* Change the [TaskArchivalPoller](https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/TaskArchiver/TaskArchiverPoller.py) to change the logic on how we clean couch records and `WMBS` records. But that would be more general change and would affect more components and would need much more testing and validation. This makes it the least favorable.
* Change the logic in [RucioInjectorPoller](https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py#L331) on how we build the list of deletable blocks and maybe add some more checks upon changing the above sql query. This to me seems the most favorable and it should be actually an addition to option 1. 

**Describe alternatives you've considered**
No alternatives

**Additional context**
None

**Comment by todor-ivanov** [comment]
Hi @hufnagel , I see you have been involved in the initial implementation of this code, and would like to ask you for any advise you may give us here. 

FYI: @drkovalskyi @germanfgv @jhonatanamado

**Comment by jhonatanamado** [comment]
Hi @todor-ivanov , thanks for document this feature. Here I can give you some hints about how the T0 agent behaves when different archivedelayhours are applied between the production and replay T0-agent shipping data to Tapes sites https://its.cern.ch/jira/browse/CMSTZ-967.

**Comment by todor-ivanov** [comment]
Thanks @todor-ivanov for giving us this feedback. This indeed is quite helpful. So this proves one possible scenario, which is:
* shortening the `archivedelayhors` won't lead to abnormal deletions.

We on the other side need also some way to answer the opposite question
* would prolonging `archivedelayhours` also drags the data being kept for longer and preventing it from been timely deleted from EOS.

**Comment by todor-ivanov** [comment]
Today we had a thorough discussion on the topic with @germanfgv (BTW Thank you very much for the extremely productive discussion German), and here are some more outcomes:

*  If we remove that [constraint](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/RucioInjector/Database/MySQL/GetDeletableBlocks.py#L53) from the `sql` query for deletable blocks to not track if workflows associated to the current block are still present in `WMBS` we will indeed (almost 99% sure now) disentangle the `archiveDelay` from `blockDeletionDelay`
* If we do so we would need another parameter to control the `blockDeletionDelay` to be put in the `RucioInjectorPoller` [deleteBlocks](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py#L331) method, which should be based now on block age rather than on workflow archival time. Because if we skip adding such a parameter, blocks of data would go to be deleted immediately once processed (closed and related transfers completed), which is not the preferable way. Some delay of this process is still needed for T0 to operate safely. 
* If we do so and base the block deletion on `block age` instead of `archival time` we will skip this piece of logic [here](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/TaskArchiver/CleanCouchPoller.py#L379-L410) from the `CleanCouchPoller`, which is actually responsible for deleting all the workflow records from `WMBS` (together with cleaning local couch etc.). So what we need to be sure about is, by changing that logic we do not overlook something that is taken into account before all the records from `WMBS` are deleted. A clue may give the set of queries used by `CleanCouchPoller` [here](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMBS/MySQL/Workflow/GetDeletableWorkflows.py), related to finding deletable workflows. 

About the block age estimation there are two possible values which could be taken into account as starting points for this clock:
* From the `dbs_buffer.block` table the `block creation time`. Since T0 blocks should be closed also by age (there should not exist an open block older than 48h - @germanfgv please correct me here if wrong) and we should be safe to estimate block age by just adding that maximum time to the `block creation time`. We should pay attention if that would be safe enough for also all blocks to be safely transferred to final destinations
* Hopefully we can find the exact moment of a block being closed and transfer completed through the `dbs_buffer` database. This to me sounds like the better way.


FYI: @drkovalskyi @amaltaro @jhonatanamado @hufnagel

**Comment by amaltaro** [comment]
If I understand this block rule deletion logic correctly, it does the following:
1. find dbs blocks where
  * the block has been closed in DBS
  * dataset has been subscribed (container level rule)
  * block is meant to be deleted (but hasn't yet)
  * the workflow has completed (`dbsbuffer_workflow.completed` is the DB hint here!!!)
2. find the expected RSE location for those blocks
3. query Rucio for block replica
4. compare if those blocks are available where they are expected to; if so, those blocks pass all the constraints for having their block-level rule deleted
5. given a list of blocks ready to have their block-level rule deleted, fetch their rules for a given account name and delete them.
6. mark these blocks as deleted in dbsbuffer table.

With all this complex - and apparently safe logic - in place, do we think that another parameter for block creation time is still required? It should be fairly simple to plug it in the system, but then we end up with yet another parameter that we need to control and keep in a reasonable value.

**Comment by todor-ivanov** [comment]
Hi @amaltaro Thanks for this addition. To me it indeed sounds as a consistent logic. There was only one part that was not projected to have while writing the above plan. And this was the Rucio set of queries. Adding another component to the system would be yet another point of failure, which for T0 should be reduced as much as possible. If we are able to collect al lthe information we need only through `dbs_buffer` database, that would be the preferable way. And actually that's why this block age estimation was suggested.  

When it comes to: 
> but then we end up with yet another parameter that we need to control and keep in a reasonable value.

And you are correct, but this is actually the searched solution. T0 team explicitly requested to have that, and they do want to have it controllable.

**Comment by amaltaro** [comment]
>  Adding another component to the system would be yet another point of failure

what is the other system that you are referring to here? If it's Rucio, then there is no way around.

> And you are correct, but this is actually the searched solution. T0 team explicitly requested to have that, and they do want to have it controllable.

Yes. However, having something requested does not mean that it's the best approach. I am not saying it is not, I am just giving extra information to be considered in this development and decisions.

**Comment by todor-ivanov** [comment]
thanks @amaltaro 

>  If it's Rucio, then there is no way around.

Yep. This is what I am talking about. Rucio is indeed already into the picture, but adding those few additional checks here, may create pressure to Rucio, which should be taken into account. I have no way to estimate in advance how big that pressure would be though.

**Comment by hufnagel** [comment]
Just read through all of this. I don't think what you propose will work, for safety and performance reasons.

The Tier0 does have the paused job concept, which means there is no predetermined time by which a workflow is done. We (in principle) can keep trying to get some workflow through for a long time (although admittedly there are practical limits in real operations). This causes problems for any age based checks. You basically would have to set the timeout so long as to never be able to impede with workflow recovery efforts and even then it's still a hard-coded magic number that needs to be set just right for the system to work correctly, which is just terrible design. Basically to have an age based check be safe you would have to do all the other checks for workflow closeout as well (that are now done before a workflow is archived). Which brings us to performance, doing these checks in two places would be expensive.

The block deletion checks are already expensive, as are workflow closeout checks. Having a lot of records in the tables involved makes these checks even more expensive, and not in a linear way, but exponentially due to the inner table joins. And these queries run over and over again, every polling cycle. You want to clean these tables asap to avoid super long query times.

**Comment by todor-ivanov** [comment]
Hi @hufnagel  Thank you very much for adding those so important details and warnings. 

I completely agree with your statements about the additional load those double checks would add to the system. As i already mentioned I had no idea how to estimate how big that load would become. 
I am going through your comment few more times now so I can put all the pieces together. But one thing is already clear, this is not going to be an easy one.

**Comment by hufnagel** [comment]
Little history, looking back on why I wrote this code the way I did. I had to make sure that workflows were not currently active in the system before allowing blocks to be deleted. And I wanted to keep the query as light as possible, given that not matter what this was going to be an expensive check that would be called every few minutes.

So I reused the outcome of the TaskArchival workflow completion check to avoid having to do all the work again in the block deletion checks.

Anyways, seems @amaltaro likes the idea of separating the archival of workflows in the agent and wmstats. Which, if implemented, would remove the need to do anything for this ticket, wouldn't it?

**Comment by amaltaro** [comment]
If I understand that SQL query correctly, I think the last two lines:
```
             HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
             AND COUNT(wmbs_workflow.name) = 0
```

are somehow redundant on what concerns workflow completion. The former ensures that that workflow has been completely processed; while the latter ensures that the workflow is no longer in the agent (aka, it's been archived).

Given that German implemented an extra check for (@germanfgv please correct me if I'm missing anything):
* container subscription; and
* block availability where the container is supposed to be

I fail to see any reason to keep the archival check in this code (the very last line in the SQL).

Please let me know what your thoughts are.

**Comment by germanfgv** [comment]
Yes @amaltaro. As discussed yesterday with @todor-ivanov, it seems like the archival check could be removed, while ensuring that the workflows are completed by checking DBSBuffer. Nonetheless, Todor commented he wants to make sure we are not accidentally skipping some important steps by disregarding the `TaskArchival` and `CouchCleanerPoller`'s logic involved in the process of removing the workflow from WMBS. 

From my side, it looks that it could just be removed, as long as we have a `RucioInjector` parameter that allows to directly set a delay on Rucio block deletions. Right now we do that indirectly via `archiveDelay`. Hence the outcomes that Todor reported [here](https://github.com/dmwm/WMCore/issues/11042#issuecomment-1070793536)

**Comment by germanfgv** [comment]
And yes, we also check for:

- Container subscription[ here](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py#L369)
- Block availability [here](https://github.com/dmwm/WMCore/blob/b77169aa015795d8725dd856f49023af9995549d/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py#L387)

The only other thing we need to make sure before deleting, is workflow completion, and it looks to me like DBS Buffer check should be enough for that.

**Comment by drkovalskyi** [comment]
Can we replace 

`COUNT(wmbs_workflow.name) = 0` 

with something that checks that we don't have workflows in any status but `completed`? If you are sure that the DBS Buffer check is sufficient, then it's not really needed.

**Comment by todor-ivanov** [comment]
> with something that checks that we don't have workflows in any status but completed

That's a good idea. 
And while looking into the options - yet again turned out not to be straight forward. Because the workflow status seems to be missing as single column at  `wmbs_workflow` table. Maybe we may get it from somewhere else. Any ideas @amaltaro @hufnagel  ?


```
MariaDB [wmagent]> describe wmbs_workflow;
+--------------+------------------+------+-----+---------+----------------+
| Field        | Type             | Null | Key | Default | Extra          |
+--------------+------------------+------+-----+---------+----------------+
| id           | int(11)          | NO   | PRI | NULL    | auto_increment |
| spec         | varchar(700)     | NO   |     | NULL    |                |
| name         | varchar(255)     | NO   | MUL | NULL    |                |
| task         | varchar(1250)    | NO   |     | NULL    |                |
| type         | varchar(255)     | YES  |     | NULL    |                |
| owner        | int(11)          | NO   | MUL | NULL    |                |
| alt_fs_close | int(1)           | NO   |     | NULL    |                |
| injected     | int(1)           | YES  |     | 0       |                |
| priority     | int(10) unsigned | YES  |     | 0       |                |
+--------------+------------------+------+-----+---------+----------------+
9 rows in set (0.003 sec)
```

**Comment by amaltaro** [comment]
There is no workflow status in wmbs_workflow. It's a table with workflow and its tasks.

The only place that I am aware of which says whether a workflow is completed or not - meaning, it processed all the input data/filesets - is in the dbsbuffer table. I am not 100% sure of this though, but 99.9% :)

**Comment by amaltaro** [comment]
> The extra line is there for a reason. And that reason is that dbsbuffer_workflow only checks the specific workflow, i.e. the reco step in PromptReco, while the join on name catches all the workflows, i.e. all the steps in PromptReco.

Just to make sure, when you say `steps` here, you are referring to different T0 workflows, right? Or is it ONE PromptReco workflow (what you see in t0_request, or what you see in t0_wmstats) with multiple `cmsRun` steps?

If the former, then yes, `dbsbuffer_workflow` will not aggregate data across workflows. It will join that block only with the workflow/task that produced it.

**Comment by hufnagel** [comment]
I mean that when Tier0Ops says workflow, they mean PromptReco for a run/dataset. Which internally in WMBS is broken up in multiples tasks, each with its own wmsb_workflow record. And these records are tied together by having the same name (hence the join on name in the query).

**Comment by hufnagel** [comment]
Basically, dbsbuffer_workflow tracks task output. It doesn't care whether that task output is used as an input for another task.

**Comment by amaltaro** [comment]
> I mean that when Tier0Ops says workflow, they mean PromptReco for a run/dataset. Which internally in WMBS is broken up in multiples tasks, each with its own wmsb_workflow record. And these records are tied together by having the same name (hence the join on name in the query).

to make sure I understand this. A T0 workflow for run 123 from dataset  MB, would be:
```
Repack_Run123_MB
Express_Run123_MB
PromptReco_Run123_MB
```
and maybe another "item" that I forgot. Is that correct? @hufnagel  (if so, then it explains why the confusion when working with central production and T0 :))

**Comment by hufnagel** [comment]
It's basically the same thing as for normal production. What Ops calls a "workflow" is what gets tracked in wmstats, i.e. it's really a request. Which is very different from what we track in the wmbs_workflow record.

**Comment by hufnagel** [comment]
For the block deletion, I basically want to keep data for a request around, as long as any element of that request is active.

PS: Additional complication is that Repack and PromptReco are separate requests, but the Repack output is the PromptReco input and this is tracked internally only through TaskArchiver workflow completion checks.

**Comment by amaltaro** [comment]
>It's basically the same thing as for normal production. What Ops calls a "workflow" is what gets tracked in wmstats, i.e. it's really a request. Which is very different from what we track in the wmbs_workflow record.

wmbs_worflow tracks workflow and its internal tasks. But it's all part of the same workflow (one can read it as a request as well). 

I am still missing this important concept of workflow (or step) in T0. Up to know, these 3 items:
```
Repack_Run123_MB
Express_Run123_MB
PromptReco_Run123_MB
```
are 3 distinct workflows. The only way I see to join those would be through input dataset and expected output dataset. 

> For the block deletion, I basically want to keep data for a request around, as long as any element of that request is active.

And what is a `request` here? If it's the workflow - similar to production - then dbsbuffer_workflow should ensure that.

**Comment by hufnagel** [comment]
> > It's basically the same thing as for normal production. What Ops calls a "workflow" is what gets tracked in wmstats, i.e. it's really a request. Which is very different from what we track in the wmbs_workflow record.
> 
> wmbs_worflow tracks workflow and its internal tasks. But it's all part of the same workflow (one can read it as a request as well).

Sure, but there are multiple wmbs_workflow table entries for the same request (same name), which makes any queries that look at wmbs_workflow more complex.

> I am still missing this important concept of workflow (or step) in T0. Up to know, these 3 items:
> 
> ```
> Repack_Run123_MB
> Express_Run123_MB
> PromptReco_Run123_MB
> ```
> 
> are 3 distinct workflows. The only way I see to join those would be through input dataset and expected output dataset.

Express_Run123_MB are standalone requests.

Repack_Run123_MB and PromptReco_Run123_MB are tied together, since the output filesets of the former are the input filesets of the latter.

> > For the block deletion, I basically want to keep data for a request around, as long as any element of that request is active.
> 
> And what is a `request` here? If it's the workflow - similar to production - then dbsbuffer_workflow should ensure that.

Yes, it would, if the requests were independent. Which they aren't due to the coupling of Repack and PromptReco.

**Comment by germanfgv** [comment]
@hufnagel please correct me if missunderstood:
1. The `COUNT(wmbs_workflow.name) = 0` line is preventing the output of Repack from being deleted before PromptReco is done.
1.From this, it is implied that Repack workflows are not removed from `wmbs_workflow` until their associated PromptReco workflow is completed.

I have not been able to find where is the logic for statement 2, that prevents workflows whose output is used by other tasks from being archived. Do you know where it is?

**Comment by todor-ivanov** [comment]
Just to have everybody on the same page, I want to visualize/summarize here the status transition/relation for t0 requests (mostly for the relation between `Repack` and `PromptReco`), as we have fully investigated with @germanfgv last week and we have concluded it should be reflecting the real picture at the moment. I'll try to be brief to the best I can.

We are currently having this picture:
```
                                    Archive Delay Period
                                  Workflows NOT archived
                            Blocks closed && NOT deleted
                                        -------->|--|<--
  +------------+                                 |  |
  | EXPRESS    |                                 |  | RepackArchived
  +------------+                                 |  |--------->
                       Repack Completed but      |  |
  +----------------+- - - - - - - - - - - - - - +|  |
  | REPACK         |       NOT archived         ||  |
  +----------------|- - - - - - - - - - - - - - +|  |
                   | Blocks closed && NOT deleted|  |       Blocks DELETED
                   |                             |  |
        |          |                             |  |
        |-->PrRNew |-->RepackInj                 |  |
        |                                        |  |
        +---------------------------------------+|  |
        | PROMPTRECO                            ||  |
        +---------------------------------------+|  |
                                                 |  |
                                                 |  | PromptRecoArchived
                                                 |  |--------->
                                                 |  |

```   

Here what we know is- we cannot inject `Repack` until `PromptReco` is released, so this way the input for `PromptReco` is protected by the lack of this flag. Once the `Repack` is injected it never goes to `normal-archived` until `PromptReco` is complete. And this is the other way how it protects the `PromptReco` input, because currently the checks are organized in the following sequence: First we do the `archival` status check (and this  way we  know there is nothing left in WMBs for the workflow in question) and only then we delete block level rules:
```
 +------------------+      +-----------------+
 | Archival Process | ---> | Block Deletion  |
 +------------------+      +-----------------+
```
Which means expanding the `Archive Delay Period` will also drag the Block level rules lifetime.

What we do want to achieve with this issue is this:

```
                                    Archive Delay Period
                                  Workflows NOT Archived
                                       && Blocks DELETED
                                        -------->|--|<--
  +------------+                                 |  |
  | EXPRESS    |                                 |  | RepackArchived
  +------------+                                 |  |--------->
                       Repack Completed but      |  |
  +----------------+- - - - - - - - - - - - - - +|  |
  | REPACK         |       NOT archived         ||  |
  +----------------|- - - - - - - - - - - - - - +|  |
                   | Blocks closed && NOT deleted|  |
                   |                             |  |
        |          |                             |  |
        |-->PrRNew |-->RepackInj                 |  |
        |                                        |  |
        +---------------------------------------+|  |
        | PROMPTRECO                            ||  |
        +---------------------------------------+|  |
                                                 |  |
                                                 |  | PromptRecoArchived
                                                 |  |--------->
                                                 |  |
```
Basically we want to start deleting blocks before the `archival` status transition. And for this we need to reverse this dependency and like that:
```
                         |            |
 +-----------------+     |            |     +------------------+
 | Block Deletion  | --> | Arch delay | --> | Archival Process |
 +-----------------+     |            |     +------------------+
                         |            |
```  
Which will cost us some additional heavy ORACLE queries in WMBS.

And just to go further to explain what exactly @hufnagel  was trying to protect us from doing was this very naive and bad idea to tie the block level deletions with some magically figured best possible  blocklevel lifetime  number. This would have brought us to dis disastrous picture:
```
                                                     Archive Delay Period
                                                   Workflows NOT Archived
                                                        && Blocks DELETED
                                                         -------->|--|<--
  +------------+                                                  |  |
  | EXPRESS    |                                 |                |  | RepackArchived
  +------------+                                 |                |  |--------->
                       Repack Completed but      |  Repack        |  |
  +----------------+- - - - - - - - - - - - - - +|  Blocks        |  |
  | REPACK         |       NOT archived         ||  DELETED       |  |
  +----------------|- - - - - - - - - - - - - - +|  Based on      |  |
                   | Blocks closed && NOT deleted|  BlockLifetime |  |
                   |                             |                |  |
        |          |                             |                |  |
        |-->PrRNew |-->RepackInj                                  |  |
        |                                                         |  |
        +--------------------------------------------------------+|  |
        | PROMPTRECO                                             ||  |
        +--------------------------------------------------------+|  |
                                                                  |  |
                                                                  |  | PromptRecoArchived
                                                                  |  |--------->
                                                                  |  |

```
In which,  inevitably sooner or later we would have ended up with a set of `PromptReco` workflows with untimely deleted input. Thanks @hufnagel for making this clear ... we would have got into trouble if you did not step in here. 


And finally just as a side comment... what we are also trying to achieve with this other issue: https://github.com/dmwm/WMCore/issues/10904  and PR: https://github.com/dmwm/WMCore/pull/11122 is this
```
                                    Archive Delay Period
                                  Workflows NOT archived
                            Blocks closed && NOT deleted
                                        -------->|--|<--
  +------------+                                 |  |
  | EXPRESS    |                                 |  | RepackArchived     |WMStats couch docs deletion
  +------------+                                 |  |--------->          |--------->
                       Repack Completed but      |  |                    |
  +----------------+- - - - - - - - - - - - - - +|  |                    |
  | REPACK         |       NOT archived         ||  |                    |
  +----------------|- - - - - - - - - - - - - - +|  |                    |
                   | Blocks closed && NOT deleted|  |                    |
                   |                             |  |                    |
        |          |                             |  |                    |
        |-->PrRNew |-->RepackInj                 |  |                    |
        |                                        |  |                    |
        +---------------------------------------+|  |                    |
        | PROMPTRECO                            ||  |                    |
        +---------------------------------------+|  |                    |
                                                 |  |                    |
                                                 |  | PromptRecoArchived |
                                                 |  |--------->          |
                                                 |  |                    |

```
Basically adding some extra `Couch Cleanup Delay` period at  the end of all this process ... so we can avoid the overload for ORACLE. But it turns out we are hitting some other CouchDB scalability constraints and we need to extra complicate the workflow state transition and cleanup process at  the end of the workflow  lifetime and possible maybe affecting also agent's management operations and procedures - needing to keep an agent  after it has been fully drained (still to be confirmed though)

We are about to put some deeper details soon on how do we exactly think stuff are happening through the full set of WMBS state transition queries. So others can also spot in case we have missed something. 

FYI: @hufnagel @amaltaro @germanfgv @jhonatanamado @drkovalskyi

**Comment by vkuznet** [comment]
@todor-ivanov first, thanks for diagrams, I think DMWM group will greatly benefit if we'll start collecting such diagrams and storing them in common area for documentation purposes. I strongly suggest that you'll convert them into some flowcharts and creates png's. Then add png's to README.md and keep in some area under WMCore.

Said that, and reading through your comments, I have doubts about your proposal. Can you clearly define `Couch Cleanup Delay`? I think you can't, is it a one minute or 1 hour? Instead, I rather see a transition flag written to DB level to indicate that, if such flag/lock exists one state can't go to another, then, along with workflow process you add function to add or release such flag/lock which will allow other processes to proceed. It would be much more reliable approach than depend on some time **undefined** time interval.

Bottom line, you can enrich your diagram with such flag/lock/semaphore which will clearly indicate when and how transition should happen.

**Comment by todor-ivanov** [comment]
> Bottom line, you can enrich your diagram with such flag/lock/semaphore which will clearly indicate when and how transition should happen.

Thanks @vkuznet  that's a good idea. I'll start accumulating those in the relevant wikies and will enrich them with all the flags and states. The current ones were just to be able to push a not very detailed visualization of the overall picture in the GH comments. But I will follow it in the documentation the proper way for sure.

As of what couch cleanup delay should mean this is easy - this should be the interval of time after we  announce the workflow as  archived and when we start deleting all related couchDB data for it in the local CouchDB instance at the agent. this one should be configurable one as explained in the other issue. But this one should be addressed in https://github.com/dmwm/WMCore/pull/11122
